File: go/transfer/ipc.go
package transfer

import (
	"fmt"
	"log/slog"
	"net"
	"os"

	"github.com/MrFaiman/uniflow/pb"
)

func StartIPCServer(sender *Sender) {
	loadDotEnv()
	path := socketPath()
	_ = os.RemoveAll(path)
	listener, err := net.Listen("unix", path)
	if err != nil {
		slog.Error("listen failed", "path", path, "err", err)
		os.Exit(1)
	}
	defer func() { _ = listener.Close() }()

	slog.Info("ipc listening", "path", path)

	for {
		conn, err := listener.Accept()
		if err != nil {
			slog.Warn("accept failed", "err", err)
			continue
		}

		go func(c net.Conn) {
			defer func() { _ = c.Close() }()
			if err := handleIPCConn(c, sender); err != nil {
				slog.Error("ipc request failed", "err", err)
			}
		}(conn)
	}
}

func handleIPCConn(conn net.Conn, sender *Sender) error {
	var req pb.IPCRequest
	if err := ReadProto(conn, &req); err != nil {
		return fmt.Errorf("read request: %w", err)
	}

	slog.Info(
		"ipc command",
		"command", req.Command,
		"target_ip", req.TargetIp,
		"object_id", req.ObjectId,
		"coordinated", req.Coordinated,
	)

	resp := &pb.IPCResponse{Success: true, Message: fmt.Sprintf("handled %s", req.Command)}

	if req.TargetIp != "" {
		if err := sender.SetTarget(req.TargetIp, req.Coordinated); err != nil {
			resp.Success = false
			resp.Message = err.Error()
		}
	}

	if resp.Success {
		if err := sender.handleIPCCommand(
			req.Command,
			req.Data,
			req.ObjectId,
			req.Coordinated,
		); err != nil {
			resp.Success = false
			resp.Message = err.Error()
		}
	}

	if err := WriteProto(conn, resp); err != nil {
		return fmt.Errorf("write response: %w", err)
	}
	return nil
}

File: python/uniflow/src/uniflow/ipc.py
import logging
import socket

from uniflow.framing import read_proto, write_proto
from uniflow.pb import message_pb2
from uniflow.utils import load_dot_env, socket_path

logger = logging.getLogger(__name__)

load_dot_env()


class Ipc:
    def __init__(self, path: str | None = None) -> None:
        self.path = path or socket_path()

    def send(
        self,
        command: str,
        data: bytes = b"",
        target_ip: str = "",
        object_id: int = 0,
        coordinated: bool = True,
        relative_path: str = "",
        dest_relative_path: str = "",
        is_directory: bool = False,
    ) -> message_pb2.IPCResponse:
        request = message_pb2.IPCRequest(
            command=command,
            data=data,
            target_ip=target_ip,
            object_id=object_id,
            coordinated=coordinated,
            relative_path=relative_path,
            dest_relative_path=dest_relative_path,
            is_directory=is_directory,
        )
        logger.info("connecting to %s", self.path)
        with socket.socket(socket.AF_UNIX, socket.SOCK_STREAM) as sock:
            sock.connect(self.path)
            logger.info("sending command %s", command)
            write_proto(sock, request)
            response = read_proto(sock, message_pb2.IPCResponse)
            logger.info(
                "received response success=%s message=%s",
                response.success,
                response.message,
            )
            return response


def default_ipc() -> Ipc:
    return Ipc()
